tools/ci_build/provider_ops_disabler.py

```diff
@@ -0,0 +1,380 @@
+
+
+def disable_ops_in_provider(operators, ep_path):
+    '''rewrite provider file to exclude unused ops'''
+
+    log.info("Rewriting {}".format(ep_path))
+    onnx_op = 'ONNX_OPERATOR_KERNEL_CLASS_NAME'
+    onnx_op_len = len(onnx_op)
+    onnx_typed_op = 'ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME'
+    onnx_typed_op_len = len(onnx_typed_op)
+    onnx_versioned_op = 'ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME'
+    onnx_versioned_op_len = len(onnx_versioned_op)
+    onnx_versioned_typed_op = 'ONNX_OPERATOR_VERSIONED_TYPED_KERNEL_CLASS_NAME'
+    onnx_versioned_typed_op_len = len(onnx_versioned_typed_op)
+    version_map = {}  # {domain:{op:[v1, v2, v3 ...]}}
+
+    def fill_version_map(op_type, opset_from, opset_to, domain):
+        '''callback func to register op in version_map'''
+
+        opset_from = int(opset_from)
+        opset_to = int(opset_to)
+
+        if domain not in version_map:
+            version_map[domain] = {}
+
+        if op_type not in version_map[domain]:
+            version_map[domain][op_type] =\
+                [opset_from, opset_to] if opset_from != opset_to else [opset_from]
+
+        else:
+            if opset_from not in version_map[domain][op_type]:
+                version_map[domain][op_type].append(opset_from)
+
+            if opset_to not in version_map[domain][op_type]:
+                version_map[domain][op_type].append(opset_to)
+
+        version_map[domain][op_type].sort()  # make sure it goes up
+        return True  # end of fill_version_map(...)
+
+    def disable_op(op_type, opset_from, opset_to, domain):
+        '''callback func to check if the op is in ops'''
+
+        def find_first_bigger(vector, elem):
+            '''return index of first element that is bigger than a'''
+            start_at = 0
+            end_at = len(vector)
+            while start_at < end_at:
+                mid_at = start_at + (end_at-start_at >> 1)
+                if vector[mid_at] > elem:
+                    end_at = mid_at
+                else:
+                    start_at = mid_at + 1
+            return end_at  # end of find_first_bigger(...)
+
+        if domain not in operators or\
+           op_type not in version_map[domain]:
+            log.info("Disable {0} {1} {2} {3}".format(domain, opset_from, opset_to, op_type))
+            return True
+
+        opset_from = int(opset_from)
+        raw_opset_to = opset_to
+        opset_to = int(opset_to)
+
+        first_bigger_at = find_first_bigger(version_map[domain][op_type], opset_to)
+        if first_bigger_at == len(version_map[domain][op_type]):
+            opset_to = 999
+        elif opset_from == opset_to or opset_to not in version_map[domain][op_type]:
+            opset_to = version_map[domain][op_type][first_bigger_at]
+
+        for opset in [str(i) for i in range(opset_from, opset_to+1)]:
+            if opset in operators[domain] and\
+               op_type in operators[domain][opset]:
+                return False
+
```

Comment:
In the original script we just stored 'from' and did the below with no binary search. What does all the extra complexity here provide as an additional benefit? There are very few versions for each op, so manually implementing binary search seems unnecessarily complicated, and 'to' (when defined) should always be one less than the 'from' of the registration for a new version

```
    if domain in domain_opset_ops:
        for opset, ops in domain_opset_ops[domain].items():
            if int(start) <= opset and int(end) >= opset and op in ops:
                enabled = True
```

Reply:
What if a model has op ABC which opset 12, while the latest version you have in registration is declared as:
7,9,ABC);
?
It should be enabled but will be disabled by the original script right?